language-plutus-core/src/Language/PlutusCore/TestSupport/Denotation.hs

```diff
@@ -0,0 +1,151 @@
+-- | This module defines tools for associating PLC terms with their corresponding
+-- Haskell values.
+
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE GADTs                     #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+module Language.PlutusCore.TestSupport.Denotation
+    ( Denotation(..)
+    , DenotationContextMember(..)
+    , DenotationContext(..)
+    , denoteVariable
+    , denoteTypedBuiltinName
+    , insertVariable
+    , insertTypedBuiltinName
+    , typedBuiltinNames
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+
+import           Data.Semigroup
+import           Data.Functor.Compose
+import           Control.Monad
+import qualified Data.ByteString.Lazy as BSL
+import           Data.GADT.Compare
+import           Data.Dependent.Map (DMap)
+import qualified Data.Dependent.Map as DMap
+
+-- | Haskell denotation of a PLC object. An object can be a 'BuiltinName' or a variable for example.
+data Denotation object size r = forall a. Denotation
```

Comment:
It would be nice to have some explanation of what these are and what they're for. I know what "denotation" is, but all the interest is in what the codomain is and why you picked it! Maybe extend the module Haddock a bit.

Reply:
I have a really good explanation of this, I think. It's not tied to our development and I can turn it into a blog post. Let's not block on this, I'll ask Manuel later whether I'm allowed to spend some time writing a blog post.